syft/execution/action.py

```diff
@@ -10,4 +20,258 @@ class Action(ABC):
+        assert actual_value is None or isinstance(
+            actual_value, expected_type
+        ), f"{field_name} must be {expected_type.__name__}, but was {type(actual_value).__name__}: {actual_value}."
+
+    # These methods must be implemented by child classes in order to return the correct type
+    # and to be detected by the serdes as serializable. They are therefore marked as abstract
+    # methods even though implementations are provided. Child classes may delegate to these
+    # implementations, but must implement their own conversions to the appropriate classes.
+
+    @staticmethod
+    @abstractmethod
+    def simplify(worker: AbstractWorker, action: "Action") -> tuple:
+        """
+        This function takes the attributes of a CommunicationAction and saves them in a tuple
+        Args:
+            worker (AbstractWorker): a reference to the worker doing the serialization
+            action (CommunicationAction): a CommunicationAction
+        Returns:
+            tuple: a tuple holding the unique attributes of the CommunicationAction
+        Examples:
+            data = simplify(worker, action)
+        """
+        return (
+            sy.serde.msgpack.serde._simplify(worker, action.name),
+            sy.serde.msgpack.serde._simplify(worker, action.target),
+            sy.serde.msgpack.serde._simplify(worker, action.args),
+            sy.serde.msgpack.serde._simplify(worker, action.kwargs),
+            sy.serde.msgpack.serde._simplify(worker, action.return_ids),
+            sy.serde.msgpack.serde._simplify(worker, action.return_value),
+        )
+
+    @staticmethod
+    @abstractmethod
+    def detail(worker: AbstractWorker, action_tuple: tuple) -> "Action":
+        """
+        This function takes the simplified tuple version of this message and converts
+        it into a CommunicationAction. The simplify() method runs the inverse of this method.
+
+        Args:
+            worker (AbstractWorker): a reference to the worker necessary for detailing. Read
+                syft/serde/serde.py for more information on why this is necessary.
+            communication_tuple (Tuple): the raw information being detailed.
+        Returns:
+            communication (CommunicationAction): a CommunicationAction.
+        Examples:
+            communication = detail(sy.local_worker, communication_tuple)
+        """
+        name, target, args_, kwargs_, return_ids, return_value = action_tuple
+
+        return (
+            sy.serde.msgpack.serde._detail(worker, name),
+            sy.serde.msgpack.serde._detail(worker, target),
+            sy.serde.msgpack.serde._detail(worker, args_),
+            sy.serde.msgpack.serde._detail(worker, kwargs_),
+            sy.serde.msgpack.serde._detail(worker, return_ids),
+            sy.serde.msgpack.serde._detail(worker, return_value),
+        )
+
+    @staticmethod
+    @abstractmethod
+    def bufferize(worker: AbstractWorker, action: "Action", protobuf_action):
```

Comment:
Python doesn't check for the method signature when you implement an abstract method? Well, this language is... surprising 😅